### PR TITLE
major mosart refactor including addition of new halo capability

### LINKED ISSUE #93: "if (ierr /= PIO_NOERR)" will not be invoked unless PIO_BCAST_ERROR is explicitly set
Unless PIO_BCAST_ERROR is explicitly set the pio error will be handled in the pio library and the code will not return to this location with an error.

_Originally posted by @jedwards4b in https://github.com/ESCOMP/MOSART/pull/76#discussion_r1438918283_

### LINKED ISSUE #97: Remove RTM in more of the MOSART code (filenames, subroutines, variables etc.)
Remove use of RTM in more of the MOSART code to distinguish it more from RTM. This is a nice cleanup step that wasn't done when MOSART started from RTM code and changed into it's own component. So something that could/should have been done earlier, but still a good thing to do now.

This is done in #76

### LINKED ISSUE #98: Change namelist items to remove "rtm" in the names of namelist variables
namelist variables have rtm in the name. This is similar to #97 except here I'm just mentioning namelist variables, which is important to note since they are using facing.

The main reason for having RTM in the names of namelist variables goes back to when RTM was a subroutine call from within CLM. So this is removing an old feature that isn't needed anymore.

This is done in #76

### LINKED ISSUE #99: Add a new mosart_noresm testlist
For use by NorESM they want to have a mosart_noresm testlist for their use on the Norwegian machine betzy.

This is done in #76

### PR DESCRIPTION
**Pull request by mvertens**
Removed all references to rtm
- files have been renamed and namelists no longer contain rtm in the name

New modularity: 
- introduced new modules with new derived types and methods
  - mosart_control_type.F90
  - mosart_tctl_type.F90
  - mosart_tparameter_type.F90
  - mosart_tspatialunit_type.F90
  - mosart_tstatusflux_type.F90
- the new modules modularize a lot of the complexity and variables that were previously found in RunOffMod.F90 and permit decomposition initialization to be more flexible and transparent.

New halo capability
- Ability to have halo regions and communication using ESMF. This is needed for computing derivatives in upcoming new additions to MOSART.
- New halo namelist - use_halo_option. When this is set to true halos can be activated. See the `test_halo` subroutine in `mosart_control_type.F90` module.
- Verified that the results for the halos are bfb identical regardless of the number of processors that are used.
-To set the values for the exclusive region that will be used in halo operations - you need to access the pointer as is done in the test_halo routine in mosart_control_type.F90:
```F90
     n = 0
      do nr = this%begr,this%endr
         n = n + 1
         this%halo_arrayptr(n) = this%latc(nr)*10. + this%lonc(nr)/100.
      end do

      call ESMF_ArrayHalo(this%haloArray, routehandle=this%haloHandle, rc=rc)
      if (chkerr(rc,__LINE__,u_FILE_u)) return

```
Issues addressed:
Fixes #93 
Fixes #98 
Fixes #97
Fixes #99

Testing on derecho:
- Verified that the following tests passed and were bfb with master
  -   ERP_Ld3.f19_f19_mtn14.2000_DATM%GSWP3v1_CLM50%SP_SICE_SOCN_MOSART_SGLC_SWAV 
  -   SMS_D_Ln5.f19_f19_mtn14.2000_DATM%GSWP3v1_CLM50%SP_SICE_SOCN_MOSART_SGLC_SWAV
   -  In both of the above use_halo_option was set to .true. in user_nl_mosart.
- Verified that answers did not change with respect to mosart1_0_49 
  -  generated baselines xml-category mosart using cesm2_3_beta17 (baseline directory is called mosart1_0_49)
  - modified cesm2_3_beta17 to use mosart branch feature/refactor and reran the mosart tests - all results were bfb

### CODE REVIEW

File: src/riverroute/mosart_physics_mod.F90 [hunk 1]
```diff
@@ -0,0 +1,604 @@
...
+
+   public :: Euler
+   public :: updatestate_hillslope
+   public :: updatestate_subnetwork
+   public :: updatestate_mainchannel
+   public :: hillsloperouting
+   public :: subnetworkrouting
+   public :: mainchannelrouting
+
+   real(r8), parameter :: TINYVALUE = 1.0e-14_r8 ! double precision variable has a significance of about 16 decimal digits
```
Comment (jedwards4b) [inline src/riverroute/mosart_physics_mod.F90:29]: use ISO_FORTRAN_ENV, only : REAL64
TINYVALUE = tiny(1._REAL64)

Comment (jedwards4b) [reply src/riverroute/mosart_physics_mod.F90:29]: This should probably be defined in shr_kind_mod.

File: src/riverroute/mosart_physics_mod.F90 [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+      !      (TRunoff%wr(nr,nt)/theDeltaT + TRunoff%dwr(nr,nt))/TRunoff%wr(nr,nt) < -TINYVALUE) then
+      !       write(iulog,*) 'negative wr!', TRunoff%wr(nr,nt), TRunoff%dwr(nr,nt), temp_dwr, temp_gwl, temp_gwl0, theDeltaT
+      !       stop
+      !    end if
+
+   end subroutine Routing_KW
+
+   !-----------------------------------------------------------------------
+
+   subroutine Routing_MC(nr, nt, theDeltaT)
```
Comment (jedwards4b) [inline src/riverroute/mosart_physics_mod.F90:348]: Should these return something about being unsupported?  Or just be removed?

File: src/riverroute/mosart_restfile.F90 [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+   use shr_kind_mod,       only : r8 => shr_kind_r8, CL => shr_kind_cl
+   use shr_sys_mod,        only : shr_sys_abort
+   use mosart_vars,        only : iulog, inst_suffix, caseid, nsrest, &
+                                  spval, mainproc, nsrContinue, nsrBranch, nsrStartup, &
+                                  ctitle, version, username, hostname, conventions, source
+   use mosart_data,        only : ctl, Trunoff
+   use mosart_histfile,    only : mosart_hist_restart
+   use mosart_fileutils,   only : getfil
+   use mosart_timemanager, only : timemgr_restart, get_nstep, get_curr_date
+   use mosart_io
```
Comment (jedwards4b) [inline src/riverroute/mosart_restfile.F90:14]: Add only clause

File: src/riverroute/mosart_restfile.F90 [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+
+      ! Local variables
+      logical :: readvar          ! determine if variable is on initial file
+      integer :: nt,nv,n          ! indices
+      real(r8) , pointer :: dfld(:) ! temporary array
+      character(len=32)  :: vname,uname
+      character(len=255) :: lname
+      !-------------------------------------
+
+      do nv = 1,7
```
Comment (jedwards4b) [inline src/riverroute/mosart_restfile.F90:366]: Remove hard coded values, use descriptive parameters.

File: src/riverroute/mosart_timemanager.F90 [hunk 5]
```diff
@@ -0,0 +1,917 @@
+module mosart_timemanager
+
+   use shr_kind_mod, only: r8 => shr_kind_r8
+   use shr_sys_mod , only: shr_sys_abort
+   use mosart_vars , only: isecspday, iulog, nsrest, nsrContinue, mpicom_rof, mainproc
+   use mosart_io
```
Comment (jedwards4b) [inline src/riverroute/mosart_timemanager.F90:6]: Add only clauses

File: src/riverroute/mosart_timemanager.F90 [hunk 6]
```diff
@@ -0,0 +1,917 @@
...
+      integer, intent(in)          :: rc   ! return code from time management library
+      character(len=*), intent(in) :: mes  ! error message
+      if ( rc == ESMF_SUCCESS ) return
+      write(iulog,*) mes
+      call shr_sys_abort ('CHKRC')
+   end subroutine chkrc
+
+   !=========================================================================================
+
+   function to_upper(str)
```
Comment (jedwards4b) [inline src/riverroute/mosart_timemanager.F90:886]: Replace with shr_string_toUpper

File: src/riverroute/mosart_tspatialunit_type.F90 [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+
+  ! Topographic and geometric properties, applicable for both grid- and subbasin-based representations
+
+   use shr_kind_mod,      only : r8 => shr_kind_r8, CL => SHR_KIND_CL
+   use shr_sys_mod,       only : shr_sys_abort
+   use shr_mpi_mod,       only : shr_mpi_sum, shr_mpi_max
+   use mosart_io,         only : ncd_pio_openfile, compDOF
+   use mosart_vars,       only : mainproc, mpicom_rof, iulog
+   use nuopc_shr_methods, only : chkerr
+   use pio
```
Comment (jedwards4b) [inline src/riverroute/mosart_tspatialunit_type.F90:11]: add only clauses

File: src/riverroute/mosart_vars.F90 [hunk 8]
```diff
@@ -0,0 +1,59 @@
+module mosart_vars
+
+   use shr_kind_mod             , only : r8 => shr_kind_r8, CL => SHR_KIND_CL
+   use shr_const_mod            , only : SHR_CONST_CDAY,SHR_CONST_REARTH
+   use shr_sys_mod              , only : shr_sys_abort
+   use ESMF
```
Comment (jedwards4b) [inline src/riverroute/mosart_vars.F90:6]: Add only clause

**Review by jedwards4b — CHANGES_REQUESTED** [review]
Throughout the code: hardcoded string lengths should be replaced with shr_kind parameters and use statements should have only clauses.   I realize that much of this code is historic, but because the filenames were changed and the history was broken future developers may not realize that.

File: cime_config/namelist_definition_mosart.xml [hunk 9]
```diff
@@ -61,30 +47,30 @@
...
     <group>mosart_inparm</group>
-    <valid_values>basin,1d,roundrobin</valid_values>
     <values>
-      <value>roundrobin</value>
+      <value>.false.</value>
     </values>
     <desc>
-      Decomposition Option for mosart
+      If true, add capability to have halo option for mosart fields.
+      In particarul these can be used to create derivatives using halo values
```
[the remark this answers is not in the document]
Comment (mvertens) [reply cime_config/namelist_definition_mosart.xml:72]: Done.

File: src/riverroute/mosart_fileutils.F90 [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+   ! !PUBLIC MEMBER FUNCTIONS:
+   public :: get_filename  !Returns filename given full pathname
+   public :: getfil        !Obtain local copy of file
+   !-----------------------------------------------------------------------
+
+contains
+
+   !-----------------------------------------------------------------------
+   character(len=256) function get_filename (fulpath)
```
[the remark this answers is not in the document]
Comment (mvertens) [reply src/riverroute/mosart_fileutils.F90]: Done

File: src/riverroute/mosart_histfile.F90 [hunk 11]
```diff
@@ -0,0 +1,1711 @@
...
+   private :: set_hist_filename         ! Determine history dataset filenames
+   private :: list_index                ! Find index of field in exclude list
+   private :: getname                   ! Retrieve name portion of input "inname"
+   private :: getflag                   ! Retrieve flag
+   private :: max_nFields               ! The max number of fields on any tape
+
+   ! !PRIVATE TYPES:
+   ! Constants
+   !
+   integer, parameter :: max_length_filename = 255 ! max length of a filename. on most linux systems this
```
[the remark this answers is not in the document]
Comment (mvertens) [reply src/riverroute/mosart_histfile.F90:84]: Done.

File: src/riverroute/mosart_io.F90 [hunk 12]
```diff
@@ -0,0 +1,1935 @@
...
+
+   ! Generic interfaces to write fields to netcdf files
+   !
+   use shr_kind_mod , only : r8 => shr_kind_r8, i8=>shr_kind_i8, shr_kind_cl, r4=>shr_kind_r4
+   use shr_sys_mod  , only : shr_sys_flush, shr_sys_abort
+   use shr_file_mod , only : shr_file_getunit, shr_file_freeunit
+   use shr_pio_mod  , only : shr_pio_getiosys, shr_pio_getiotype, shr_pio_getioformat
+   use mosart_vars  , only : spval, ispval, iulog, mainproc, mpicom_rof, iam, npes
+   use perf_mod     , only : t_startf, t_stopf
+   use pio
```
[the remark this answers is not in the document]
Comment (mvertens) [reply src/riverroute/mosart_io.F90]: Done.

File: src/riverroute/mosart_histfile.F90 [hunk 13]
```diff
@@ -0,0 +1,1711 @@
...
+                                  get_calendar, NO_LEAP_C, GREGORIAN_C
+   use mosart_io
+
+   implicit none
+   private
+
+   ! Constants
+   integer , public, parameter :: max_tapes  = 3     ! max number of history tapes
+   integer , public, parameter :: max_flds   = 1500  ! max number of history fields
+   integer , public, parameter :: max_namlen = 32    ! maximum number of characters for field name
```
[the remark this answers is not in the document]
Comment (mvertens) [reply src/riverroute/mosart_histfile.F90]: Done  (for max_namlen)

File: src/riverroute/mosart_restfile.F90 (hunk 4 above)
Comment (mvertens) [reply src/riverroute/mosart_restfile.F90]: Done.

File: src/riverroute/mosart_timemanager.F90 (hunk 5 above)
Comment (mvertens) [reply src/riverroute/mosart_timemanager.F90]: Done.

File: src/riverroute/mosart_timemanager.F90 (hunk 6 above)
Comment (mvertens) [reply src/riverroute/mosart_timemanager.F90]: Done.

File: src/riverroute/mosart_tspatialunit_type.F90 (hunk 7 above)
Comment (mvertens) [reply src/riverroute/mosart_tspatialunit_type.F90]: Done.

File: src/riverroute/mosart_vars.F90 (hunk 8 above)
Comment (mvertens) [reply src/riverroute/mosart_vars.F90]: Done.

File: src/riverroute/mosart_histfile.F90 (hunk 11 above)
[the remark this answers is not in the document]
Comment (jedwards4b) [reply src/riverroute/mosart_histfile.F90:84]: max_length_filename = CL

[the remark this answers is not in the document]
Comment (mvertens) [reply src/riverroute/mosart_histfile.F90]: Done.

File: src/riverroute/mosart_restfile.F90 (hunk 4 above)
Comment (jedwards4b) [reply src/riverroute/mosart_restfile.F90]: not done?

**Review by jedwards4b — APPROVED** [review]
This is an excellent refactor.  Thank you so much.

**Comment by ekluzek** [comment]
@mvertens can you tell us more about the new capability that is planned to MOSART after this refactor. For CESM4 we are hoping that we can deprecate and eliminate the use of MOSART in CESM and ONLY use mizuRoute.

A big limitation of MOSART is that we don't have tools to make new input datasets nor people we can rely on to make the datasets. This means MOSART can NOT be used for Paleo work for example, which kills it's usage for CESM4. At NCAR we don't have local experts for MOSART only for mizuRoute. Removing RTM and MOSART for CESM4 would allow us to only support one river model for CESM4 which would lower the projects maintenance costs.

Have you looked into adding these same changes to mizuRoute instead of (or in addition to) MOSART? We should at least look into that.

**Comment by mvertens** [comment]
@ekluzek - the key capability that was needed is halo updates. This is currently needed by @swensosc for his work and he was targeting MOSART. I think doing this in MOSART was very easy and MOSART is the code base that will be supported in both NorESM3 and CESM3. So having a clean refactored code will be helpful to both efforts in my opinion. As for MizuRoute - I'm happy to talk about that in the future.

**Comment by ekluzek** [comment]
I think this is a big enough of a change that we should bump the middle number in the version sequence.
And we might as well make the name more similar to CTSM tags.

So this becomes

mosart1.1.dev01

**Comment by jedwards4b** [comment]
I have to say - I don't like the 'dev' part.  Just make it mosart1.1.01

**Comment by ekluzek** [comment]
In talking about this today, there's one bit of additional functionality that will be added (having to do with gradients). This tag will be the basis for the GW branch, as well as the DOM work that is being done at NIO.

We also decided that this would be done after the list of simple PR's that will be done first.

**Comment by mvertens** [comment]
@slevis-lmwg @ekluzek @jedwards4b - I have verified that this branch does not change answers with respect to mosart1_0_49. @mvdebolskiy also participated in updating this refactor branch to have updated internal diagnostics. This branch will be the starting point of the new glc->rof coupling that will replace glc->ocn.

**Comment by ekluzek** [comment]
The plan right now is for @slevis-lmwg to bring this in after he has some time clear with bringing CN Matrix to CTSM.

File: src/riverroute/mosart_io.F90 [hunk 14]
```diff
@@ -0,0 +1,1935 @@
...
+      character(len=*) , intent(in)    :: fname  ! Input filename to open
+      integer          , intent(in)    :: mode   ! file mode
+      ! !LOCAL VARIABLES:
+      integer :: ierr
+      character(len=*),parameter :: subname='ncd_pio_openfile' ! subroutine name
+      !-----------------------------------------------------------------------
+
+      ierr = pio_openfile(pio_subsystem, file, io_type, fname, mode)
+
+      if(ierr/= PIO_NOERR) then
```
[the remark this answers is not in the document]
Comment (slevis-lmwg) [reply src/riverroute/mosart_io.F90:205]: opened issue for this

File: src/riverroute/mosart_io.F90 [hunk 15]
```diff
@@ -0,0 +1,1935 @@
...
+      !-----------------------------------------------------------------------
+
+
+      if ( present(print_err) )then
+         log_err = print_err
+      else
+         log_err = .true.
+      end if
+      readvar = .true.
+      call pio_seterrorhandling(ncid, PIO_BCAST_ERROR)
```
[the remark this answers is not in the document]
Comment (slevis-lmwg) [reply src/riverroute/mosart_io.F90:287]: Linked from the same issue that I opened for the last comment.

**Comment by slevis-lmwg** [comment]
OK derecho test-suite
FAIL izumi test-suite during build; I have restarted the build in case it was a glitch; see here in case anyone wishes to take a look before Erik and I get to it:
`/fs/cgd/data0/slevis/git_externals/major_refactor/tests_0605-174443iz`

**Comment by mvertens** [comment]
@slevis-lmwg - there were indeed compilation problem with nag for this PR. I have fixed the problems and confirmed that mosart can now build with nag on izumi. I have pushed the changes back.

**Comment by mvertens** [comment]
@slevis-lmwg - I have addressed #93 and pushed the changes back.

File: src/riverroute/mosart_mod.F90 [hunk 16]
```diff
@@ -0,0 +1,983 @@
...
+                                   ndens, mfilt, nhtfrq, avgflag_pertape, avgflag_pertape, &
+                                   fincl1, fincl2, fincl3, fexcl1, fexcl2, fexcl3, max_tapes, max_namlen
+   use mosart_restfile    , only : mosart_rest_timemanager, mosart_rest_getfile, mosart_rest_fileread, &
+                                   mosart_rest_filewrite, mosart_rest_filename, finidat, nrevsn
+   use mosart_physics_mod , only : updatestate_hillslope, updatestate_subnetwork, updatestate_mainchannel, Euler
+   use perf_mod           , only : t_startf, t_stopf
+   use nuopc_shr_methods  , only : chkerr
+   use ESMF               , only : ESMF_SUCCESS, ESMF_FieldGet, ESMF_FieldSMMStore, ESMF_FieldSMM, &
+                                   ESMF_TERMORDER_SRCSEQ, ESMF_Mesh
+   use mosart_IO ! TODO: put in only here
```
[the remark this answers is not in the document]
Comment (slevis-lmwg) [reply src/riverroute/mosart_mod.F90]: @mvertens while we wait to resolve #93, let's address the remaining unresolved comments in the PR. 

"Add only clauses" here and elsewhere. Do you want to address now or open an issue?

File: src/riverroute/mosart_mod.F90 [hunk 17]
```diff
@@ -0,0 +1,983 @@
...
+      if (tod == 0) budget_check = .true.
+      budget_terms = 0._r8
+
+      ! BUDGET
+      ! BUDGET terms 1-10 are for volumes (m3)
+      ! BUDGET terms 11-30 are for flows (m3/s)
+      call t_startf('mosartr_budget')
+      do nt = 1,ntracers
+         do nr = begr,endr
+            budget_terms( 1,nt) = budget_terms( 1,nt) + ctl%volr(nr,nt)
```
[the remark this answers is not in the document]
Comment (slevis-lmwg) [reply src/riverroute/mosart_mod.F90]: @mvertens address now or open an issue?

File: src/riverroute/mosart_physics_mod.F90 (hunk 1 above)
Comment (slevis-lmwg) [reply src/riverroute/mosart_physics_mod.F90]: address now or open an issue?

File: src/riverroute/mosart_physics_mod.F90 (hunk 2 above)
Comment (slevis-lmwg) [reply src/riverroute/mosart_physics_mod.F90]: address now or open an issue?

File: src/riverroute/mosart_restfile.F90 (hunk 3 above)
Comment (slevis-lmwg) [reply src/riverroute/mosart_restfile.F90]: Same as https://github.com/ESCOMP/MOSART/pull/76#discussion_r1629728441

File: src/riverroute/mosart_mod.F90 (hunk 17 above)
[the remark this answers is not in the document]
Comment (mvertens) [reply src/riverroute/mosart_mod.F90]: @slevis-lmwg - can you please open an issue and I'll address it in the next PR that has the glc->ocn via mosart.

File: src/riverroute/mosart_physics_mod.F90 (hunk 2 above)
Comment (mvertens) [reply src/riverroute/mosart_physics_mod.F90]: Please open an issue as well.

File: src/riverroute/mosart_restfile.F90 (hunk 3 above)
Comment (mvertens) [reply src/riverroute/mosart_restfile.F90]: This has been resolved.

File: src/riverroute/mosart_mod.F90 (hunk 16 above)
[the remark this answers is not in the document]
Comment (slevis-lmwg) [reply src/riverroute/mosart_mod.F90]: mosart_mod.F90 seems gone, so I'm resolving this one as unnecessary.

**Comment by ekluzek** [comment]
Awesome, thanks for working on that @mvertens and being able to get it done quickly. We opened the issue in case this was a longer term thing that didn't need to be handled now. We figured it wasn't required for the Science capability freeze timeline, so a longer term issue also made sense. But, even better that you've already resolved it.

**Comment by slevis-lmwg** [comment]
I reran the mosart test-suite:
OK on izumi, only NLCOMP diffs
OK on derecho, only NLCOMP and an EXPECTED FAILURE

**Review by ekluzek — APPROVED** [review]
@mvertens thanks for your work on all of this. These are great improvements. Nice, to see that all of this is bit-for-bit. I didn't go over this as extensively as @jedwards4b and we are grateful for his careful feedback here.

I looked at some big picture things that were happening here, and made issues to document what was done. In the future if you can make issues on these type of refactor type things you are doing that's helpful for us to discuss them as a group before they are finished and the code and testing completed. It also gives us pre-warning of work that is happening. It's unlikely to collide with something we are doing, but really helpful for discussion and planning.

Thanks for the work here and in the next PR

**Comment by jedwards4b** [comment]
@slevis-lmwg This repository does not have automated tag creation - will you make a new tag?

**Comment by slevis-lmwg** [comment]
I type `git tag` from my latest master (local), and I get `mosart1.1.01`
which is the annotated that I created and pushed to escomp.

On github I see the same tag here:
https://github.com/ESCOMP/MOSART/tags

Please let me know if I'm not answering your question.

**Comment by jedwards4b** [comment]
I think that you must have made the tag shortly after I posted that message - thanks.

File: src/riverroute/mosart_physics_mod.F90 (hunk 2 above)
Comment (mvertens) [reply src/riverroute/mosart_physics_mod.F90]: These actually have already been removed.

File: src/riverroute/mosart_physics_mod.F90 (hunk 1 above)
Comment (jedwards4b) [reply src/riverroute/mosart_physics_mod.F90]: I have a lot on my plate right now, but if you want to open a PR to cesm_share I will happily review it.

**Pull request merged**